app/controllers/supplementary/AdditionalInformationController.scala

```diff
@@ -34,27 +39,160 @@ class AdditionalInformationController @Inject()(
   appConfig: AppConfig,
   override val messagesApi: MessagesApi,
   authenticate: AuthAction,
+  errorHandler: ErrorHandler,
   customsCacheService: CustomsCacheService
 )(implicit ec: ExecutionContext)
     extends FrontendController with I18nSupport {
 
   def displayForm(): Action[AnyContent] = authenticate.async { implicit request =>
-    customsCacheService.fetchAndGetEntry[AdditionalInformation](supplementaryCacheId, formId).map {
-      case Some(data) => Ok(additional_information(appConfig, form.fill(data)))
-      case _          => Ok(additional_information(appConfig, form))
+    customsCacheService.fetchAndGetEntry[AdditionalInformationData](supplementaryCacheId, formId).map {
+      case Some(data) => Ok(additional_information(appConfig, form, data.items))
+      case _          => Ok(additional_information(appConfig, form, Seq()))
     }
   }
 
   def saveAdditionalInfo(): Action[AnyContent] = authenticate.async { implicit request =>
-    form
-      .bindFromRequest()
-      .fold(
-        (formWithErrors: Form[AdditionalInformation]) =>
-          Future.successful(Ok(additional_information(appConfig, formWithErrors))),
-        validForm =>
-          customsCacheService.cache[AdditionalInformation](supplementaryCacheId, formId, validForm).map { _ =>
-            Redirect(controllers.supplementary.routes.DocumentsProducedController.displayForm())
+    val boundForm = form.bindFromRequest()
+
+    val actionType =
+      request.body.asFormUrlEncoded.flatMap(_.get("action")).flatMap(_.headOption).getOrElse("Wrong action")
+
+    val cachedData = customsCacheService
+      .fetchAndGetEntry[AdditionalInformationData](supplementaryCacheId, formId)
+      .map(_.getOrElse(AdditionalInformationData(Seq())))
+
+    cachedData.flatMap { cache =>
+      boundForm
+        .fold(
+          (formWithErrors: Form[AdditionalInformation]) =>
+            Future.successful(BadRequest(additional_information(appConfig, formWithErrors, cache.items))),
+          validForm =>
+            actionType match {
+              case "Add"                             => addItem(validForm, cache)
+              case "Save and continue"               => saveAndContinue(validForm, cache)
+              case value if value.contains("Remove") => removeItem(retrieveItem(actionType), cache)
+              case _                                 => errorHandler.displayErrorPage()
+          }
+        )
+    }
+  }
+
+  private def saveAndContinue(
+    userInput: AdditionalInformation,
+    cachedData: AdditionalInformationData
+  )(implicit request: AuthenticatedRequest[_], hc: HeaderCarrier): Future[Result] =
+    (userInput, cachedData.items) match {
+      case (item, Seq()) => handleSaveAndContinueEmptyCache(userInput, item)
+      case (item, items) => handleSaveAndContinueCache(userInput, item, items)
```

Comment:
`userInput` and `item` are totally the same, but different name, something goes wrong here.  Please remove one of them

Reply:
Thank you for spotting that @PatrykRudnicki .Fixed!